### ISSUE TITLE
UAS Tugas 15 Machine Learning Nessa

### ISSUE THREAD

**Issue by nessa0712** [nessa0712/UAS-Machine-Learning_Nessa_InformatikaB#5]
**Tugas 15**

1.      Buat folder baru di dalam folder python di file directory

![Image](https://github.com/user-attachments/assets/5964daa4-504b-4639-9588-005358fc14d4)

![Image](https://github.com/user-attachments/assets/eaa0a2be-b801-4d79-ba42-199a468c5ca5)
2.	Buka Cmd masuk ke directory yang sudah dibuat sebelumnya

![Image](https://github.com/user-attachments/assets/bcf6d745-b1d4-4d5e-ab57-db6aee91f38d)
3.     Install virtualenv, jika sudah ada maka bisa langsung panggil virtualenv di cmd

![Image](https://github.com/user-attachments/assets/9d7df1f1-afa1-414e-8f58-fd7991f55e51)
4.     Kemudian, aktifkan environment di env dengan scripts activate.bat

![Image](https://github.com/user-attachments/assets/7f77fa8e-9e25-4504-b313-7bd8cf7fc35d)
5.	Install flashk sqlalchemy agar dapat membuat aplikasi web

![Image](https://github.com/user-attachments/assets/1ac6d01f-31f5-41c6-b513-2592f4db326a)
6.	Install library scikit-learn 

![Image](https://github.com/user-attachments/assets/a821ba89-bc57-4429-8853-389d976593e5)
7.	Buka vscode dan buat file baru bernama ’app.py’ di dalam folder plmdiabetes
# Mencetak nama dan NIM
print('Nessa Anindia Putri')
print('41155050210057')

# Mengimpor library yang diperlukan
from flask import Flask, render_template, request, redirect  # Flask untuk membuat aplikasi web
import pickle  # Untuk memuat model yang disimpan
import sklearn  # Library machine learning
import numpy as np  # Untuk komputasi numerik
import os  # Untuk mengelola path file

# Membuat instance Flask
app = Flask(__name__)

# Mendefinisikan route utama ('/') dengan metode POST dan GET
@app.route('/', methods=['POST', 'GET'])
def index():
    # Jika metode request adalah POST (form dikirim)
    if request.method == 'POST':
        # Tentukan path lengkap ke file knn_pickle
        # os.path.join digunakan untuk menggabungkan direktori saat ini dengan nama file
        pickle_file_path = os.path.join(os.path.dirname(__file__), 'knn_pickle')
        
        # Memuat model KNN dari file knn_pickle
        with open(pickle_file_path, 'rb') as r:
            model = pickle.load(r)

        # Mengambil input dari form HTML
        melahirkan = float(request.form['melahirkan'])  # Jumlah melahirkan
        glukosa    = float(request.form['glukosa'])    # Kadar glukosa
        darah      = float(request.form['darah'])      # Tekanan darah
        kulit      = float(request.form['kulit'])      # Ketebalan kulit
        insulin    = float(request.form['insulin'])    # Kadar insulin
        bmi        = float(request.form['bmi'])        # Indeks massa tubuh (BMI)
        riwayat    = float(request.form['riwayat'])    # Riwayat diabetes dalam keluarga
        umur       = float(request.form['umur'])       # Usia

        # Menyiapkan data untuk prediksi
        # Mengubah input menjadi array numpy
        datas = np.array((melahirkan, glukosa, darah, kulit, insulin, bmi, riwayat, umur))
        # Mengubah bentuk array menjadi (1, 8) karena model mengharapkan input 2D
        datas = np.reshape(datas, (1, -1))

        # Melakukan prediksi menggunakan model
        isDiabetes = model.predict(datas)

        # Render template hasil.html dan kirim hasil prediksi (isDiabetes) ke template
        return render_template('hasil.html', finalData=isDiabetes)
    
    # Jika metode request adalah GET (halaman dimuat pertama kali)
    else:
        # Render template index.html (form input)
        return render_template('index.html')
    
# Menjalankan aplikasi Flask
if __name__ == "__main__":
    app.run(debug=True)  # debug=True untuk mode pengembangan
8.	Panggil requiretment di cmd 

![Image](https://github.com/user-attachments/assets/96e9a38f-0f21-40f3-9434-960fb3d9302e)

![Image](https://github.com/user-attachments/assets/5d4b7bad-d48c-4c59-ae0a-f3bdd07248bb)
9.	Download template web [GitHub - heriistantoo/flaskdiabet: Subscribe paperlimamenit](https://github.com/heriistantoo/flaskdiabet) lalu simpan folder css di folder static

![Image](https://github.com/user-attachments/assets/338e93b9-beb6-4e7a-92dd-2b4d893cd789)
•	Index. html 
{% extends 'base.html' %}

{% block head %}
<title>Tugas Pertemuan 15 - Machine Learning</title>
{% endblock %}

{% block body %}
<div class="container-contact100">
    <div class="wrap-contact100">
        <span class="label-input100">
            Nama : Nessa Anindia Putri Siregar
            <br>
            &nbsp; NPM  &nbsp;&nbsp;&nbsp;&nbsp;: 41155050210057
        </span>
        <form class="contact100-form validate-form" action="/" method="POST">
            
            <span class="contact100-form-title">
                Deteksi Diabetes!
            </span>

            <div class="wrap-input100 validate-input">
                <span class="label-input100">Banyak Melahirkan</span>
                <input class="input100" type="text" placeholder="Masukan jumlah melahirkan" name="melahirkan" id="melahirkan" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input" >
                <span class="label-input100">Kadar Glukosa</span>
                <input class="input100" type="text" placeholder="Masukan kadar glukosa" name="glukosa" id="glukosa" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input" >
                <span class="label-input100">Tekanan Darah</span>
                <input class="input100" type="text" placeholder="Masukan tekanan darah" name="darah" id="darah" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input" >
                <span class="label-input100">Tebal Kulit</span>
                <input class="input100" type="text" placeholder="Masukan ketebalan kulit" name="kulit" id="kulit" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input" >
                <span class="label-input100">Kadar Insulin</span>
                <input class="input100" type="text" placeholder="Masukan kadar insulin" name="insulin" id="insulin" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input" >
                <span class="label-input100">BMI</span>
                <input class="input100" type="text" placeholder="Masukan BMI" name="bmi" id="bmi" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input" >
                <span class="label-input100">Riwayat Diabetes</span>
                <input class="input100" type="text" placeholder="Masukan derajat diabetes keturunan" name="riwayat" id="riwayat" required>
                <span class="focus-input100"></span>
            </div>

            <div class="wrap-input100 validate-input">
                <span class="label-input100">Umur</span>
                <input class="input100" type="text" placeholder="Masukan umur" name="umur" id="umur" required>
                <span class="focus-input100"></span>
            </div>

            <div class="container-contact100-form-btn">
                <div class="wrap-contact100-form-btn">
                    <div class="contact100-form-bgbtn"></div>
                    <button class="contact100-form-btn">
                        <span>
                            Prediksi
                            <i class="fa fa-long-arrow-right m-l-7" aria-hidden="true"></i>
                        </span>
                    </button>
                </div>
            </div>

        </form>
    </div>
</div>

{% endblock %}

•	Hasil.html
{% extends 'base.html' %}

{% block head %}
<title>Tugas Pertemuan 15 - Machine Learning</title>
{% endblock %}

{% block body %}
<div class="container-contact100">
    <div class="wrap-contact100">
        <span class="label-input100">
            Nama : Nessa Anindia Putri Siregar
            <br>
            &nbsp; NPM  &nbsp;&nbsp;&nbsp;&nbsp;: 41155050210057
        </span>
        <form class="contact100-form validate-form" action="/" method="GET">
            {% if finalData == 1 %}
            <span class="contact100-form-title">
                POSITIF DIABETES
            </span>

            {% else %}
            <span class="contact100-form-title">
                NEGATIF DIABETES
            </span>

            {% endif %}

            <div class="container-contact100-form-btn">
                <div class="wrap-contact100-form-btn">
                    <div class="contact100-form-bgbtn"></div>
                    <button class="contact100-form-btn">
                        <span>
                            Selesai
                        </span>
                    </button>
                </div>
            </div>
            
        </form>
    </div>
</div>

{% endblock %}

•	Base.html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <link rel="icon" type="image/png" href="{{ url_for('static', filename='css/images/icons/favicon.ico') }}"/>
    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/vendor/bootstrap/css/bootstrap.min.css') }}">
    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/fonts/font-awesome-4.7.0/css/font-awesome.min.css') }}" >
    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/css/main.css') }}">

    {% block head %}{% endblock %}
</head>

<body>
    {% block body %}{% endblock %}
</body>

</html>
10.	Download knn pickle untuk melakukan serialisasi dan deserialisasi tipe data dan file pendukung lainnya di github lalu disimpan di folder project plmdiabetes

![Image](https://github.com/user-attachments/assets/98e2ecab-7aab-48e3-b7e5-f0935adcd1c1)
11.	Jalankan app.py di cmd

![Image](https://github.com/user-attachments/assets/916114f8-dff8-4f41-b228-12441dabbd58)
12.	Jalankan web di http://localhost:5000/, hasil didapatkan dari prediksi knn pickle

![Image](https://github.com/user-attachments/assets/98cb5abd-f21e-4613-915e-7c3e8eaa9490)

![Image](https://github.com/user-attachments/assets/f10d375f-5c1b-410c-8d9b-f3860b2bae11)
**Hosting**
1.	Pertama, buka python anywhere dan buatlah akun di python anywhere

![Image](https://github.com/user-attachments/assets/5a7c208a-276b-44c2-8fde-4cf2b180d1b7)
2.	Klik menu web app, kemudian add new web app 

![Image](https://github.com/user-attachments/assets/e33db776-4a28-43e2-87d6-27a2e928e7c4)
3.	Jadikan folder plmdiabetes sebagai zip dan upload ke folder mysite/

![Image](https://github.com/user-attachments/assets/15b2cfa0-31e3-4d96-8ed9-0776fa0d021f)

![Image](https://github.com/user-attachments/assets/256339c5-4527-496b-bad4-340d80756dcb)
4.	Lalu kembali ke dashboard, klik bash pada newe console

![Image](https://github.com/user-attachments/assets/dc9f1cef-9377-4005-bfd4-0753fea1ca5a)
5.     Pada terminal bash masukan kode sebagai berikut:

![Image](https://github.com/user-attachments/assets/9647fec7-8865-48a5-ab89-c95d9ab26ca5)

![Image](https://github.com/user-attachments/assets/aa6141ea-6aea-458f-b288-4a828743a153)

![Image](https://github.com/user-attachments/assets/426d897c-d7a3-4dc9-a906-b74bddbaafc3)
6.	Kemudian, pada halaman web di dashboard scroll kebawah hingga menemukan bagian code dan klik pada bagian WSGI

![Image](https://github.com/user-attachments/assets/47eb612a-b586-414f-8e51-9ce25a093d8f)
7.	Ganti flash menjadi app

![Image](https://github.com/user-attachments/assets/d81780e3-b668-4c2d-8ca8-84779c023ed4)
8.	Lalu masuk ke virtualenv dan masukan lokasi env yang sudah dibuat sebelumnya

![Image](https://github.com/user-attachments/assets/3e596cb2-ebb4-4504-8907-78e1c89d0fee)
9.	Jika sudah selesai klik reload 

![Image](https://github.com/user-attachments/assets/d0d7a3c0-4a85-4b17-a0aa-c6caf474dcac)
10.   Test halaman nessaputri.pythonanywhere.com

![Image](https://github.com/user-attachments/assets/19bc5f6d-3b08-406c-85d9-c8a06934fc5f)

![Image](https://github.com/user-attachments/assets/523f1d4f-2d41-48bc-a929-a7e799733dd5)
